### PR TITLE
feat(ingress gateway: support configuring limits in ingress-gateway c…

### PR DESCRIPTION
**Pull request by huikang**
### Description
As a user of ingress gateway, I want to increase the max_connections of the service in the ingress-gateway config entry ([link](https://github.com/hashicorp/consul/issues/13374) to the detailed description). However, in current implementation, max_coonections can’t be changed and is set to the default value of 1024 by envoy.

This PR adds a new field named Defaults to the ingress-gateway config entry and `max_xyz` to individual listening services. The new fields allow user to configure the `max_xyz`  of the upstream cluster of a specified service behind the ingress gateway.

- updated unit test and integration test
- updated external-facing doc

### Testing & Reproduction steps

- A full example is verified in the integration test (`test/integration/connect/envoy/case-ingress-gateway-multiple-services/config_entries.hcl`).

```hcl
Kind = "ingress-gateway"
Name = "my-ingress-gateway"

Defaults {
    MaxConnections = 2048          // <----- increase the max_connections of all upstream service instances
}
Listeners = [
 {
   Port = 8082
   Protocol = "http"
   Services = [
     {
        Name = "counting"
        Hosts = ["*"]
        MaxConnections = 5120          // <----- increase the max_connections, overwrite the value in Defaults
     },
     {
       Name = "counting-v2"
       Hosts = ["v2"]
     }
   ]
 }
]

```

- envoy cluster
```
counting.default.dc2.internal.ca9229ae-1b1d-af9d-47fa-43885f09557a.consul::default_priority::max_connections::5120
counting.default.dc2.internal.ca9229ae-1b1d-af9d-47fa-43885f09557a.consul::default_priority::max_pending_requests::1024
counting.default.dc2.internal.ca9229ae-1b1d-af9d-47fa-43885f09557a.consul::default_priority::max_requests::1024
counting.default.dc2.internal.ca9229ae-1b1d-af9d-47fa-43885f09557a.consul::default_priority::max_retries::3
```

### Links

Address #13374
Supersede #14708 

Please be mindful not to leak any customer or confidential information. HashiCorp employees may want to use our internal URL shortener to obfuscate links.

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [x] not a security concern

### CODE REVIEW

File: api/config_entry_gateways.go [hunk 1]
```diff
@@ -37,6 +40,12 @@ type IngressGatewayConfigEntry struct {
 	ModifyIndex uint64
 }
 
+type IngressServiceConfig struct {
+	MaxConnections        *int
+	MaxPendingRequests    *int
+	MaxConcurrentRequests *int
```
Comment (DanStough) [inline api/config_entry_gateways.go:46]: Any reason these are `int` instead of `uint32`?

File: agent/structs/config_entry_gateways.go [hunk 2]
```diff
@@ -31,11 +31,20 @@ type IngressGatewayConfigEntry struct {
...
+
 	Meta               map[string]string `json:",omitempty"`
 	acl.EnterpriseMeta `hcl:",squash" mapstructure:",squash"`
 	RaftIndex
 }
 
+type IngressServiceConfig struct {
+	MaxConnections        int32
+	MaxPendingRequests    int32
+	MaxConcurrentRequests int32
```
Comment (DanStough) [inline agent/structs/config_entry_gateways.go:45]: `uint32` would be more accurate per envoy, correct?

File: api/config_entry_gateways_test.go [hunk 3]
```diff
@@ -163,11 +170,20 @@ func TestAPI_ConfigEntries_IngressGateway(t *testing.T) {
 			readIngress.Listeners[0].Services[0].Partition = ""
 
 			require.Equal(t, ingress1.Listeners, readIngress.Listeners)
+			require.Equal(t, ingress1.Listeners[0].Services[0].MaxConnections,
+				readIngress.Listeners[0].Services[0].MaxConnections)
+			require.Equal(t, ingress1.Listeners[0].Services[0].MaxConcurrentRequests,
+				readIngress.Listeners[0].Services[0].MaxConcurrentRequests)
+			require.Equal(t, ingress1.Listeners[0].Services[0].MaxPendingRequests,
+				readIngress.Listeners[0].Services[0].MaxPendingRequests)
```
Comment (DanStough) [inline api/config_entry_gateways_test.go:178]: My understanding of `require.Equal` is that it's a deep comparison by value, so I think the new lines here are redundant.

**Review by DanStough — COMMENTED** [review]
I had some editorial comments we could discuss but overall looks great!

File: api/config_entry_gateways.go (hunk 1 above)
Comment (huikang) [reply api/config_entry_gateways.go]: Fixed. Thanks.

File: website/content/docs/connect/config-entries/ingress-gateway.mdx [hunk 4]
```diff
@@ -1041,6 +1046,35 @@ You can specify the following parameters to configure ingress gateway configurat
...
+              be respected, a L7 protocol must be defined in the \`protocol\` field.`,
+        },
+        {
+          name: 'MaxConcurrentRequests',
+          type: 'int: 0',
+          description: `The maximum number of concurrent requests that
+              will be allowed at a single point in time. Use this to limit HTTP/2 traffic,
+              since HTTP/2 has many requests per connection. For this configuration to be
+              respected, a L7 protocol must be defined in the \`protocol\` field.`,
+        },
```
[the remark this answers is not in the document]
Comment (huikang) [reply website/content/docs/connect/config-entries/ingress-gateway.mdx]: Added in the following commit

File: website/content/docs/connect/config-entries/ingress-gateway.mdx [hunk 5]
```diff
@@ -1156,6 +1190,21 @@ You can specify the following parameters to configure ingress gateway configurat
                 },
               ],
             },
+            {
+              name: 'MaxConnections',
+              type: 'int: 0',
+              description: overrides for `Defaults` configuration,
```
[the remark this answers is not in the document]
Comment (huikang) [reply website/content/docs/connect/config-entries/ingress-gateway.mdx]: Added in the following commit

File: agent/structs/config_entry_gateways.go [hunk 6]
```diff
@@ -31,11 +31,20 @@ type IngressGatewayConfigEntry struct {
 	// what services to associated to those ports.
 	Listeners []IngressListener
 
+	// Defaults contains default configuration for all upstream service instances
+	Defaults *IngressServiceConfig `json:",omitempty"`
+
```
[the remark this answers is not in the document]
Comment (DanStough) [reply agent/structs/config_entry_gateways.go:36]: Per our discussion, it seems like it's fine where it is 👍

**Review by DanStough — APPROVED** [review]
Awesome! 🚀

**Comment by david-yu** [comment]
@huikang This will also need a PR on consul-k8s to apply this new config via CRDs

**Pull request merged**